### PR TITLE
use release version of pdf.js

### PR DESCRIPTION
**Pull request by liudongmiao**
As discussing in https://github.com/FirstTimeInForever/intellij-pdf-viewer/pull/114, there are several options to include pdf.js.
In https://github.com/FirstTimeInForever/intellij-pdf-viewer/pull/114, we choose compile from pdf.js's source code.

However, we don't need to.

This PR simplify the processes, just download and unzip, the inject intellij-pdf-viewer's js and css, make it much easy.

### CODE REVIEW

File: plugin/src/main/kotlin/com/firsttimeinforever/intellij/pdf/viewer/jcef/PdfStaticServer.kt [hunk 1]
```diff
@@ -112,11 +112,11 @@ internal class PdfStaticServer : HttpRequestHandler() {
   fun getPreviewUrl(file: VirtualFile, withReloadSalt: Boolean = false): String {
     val salt = if (withReloadSalt) Random.nextInt() else 0
     vfsMap[file.url] = file
-    val url = parseEncodedPath("$serverUrl/index.html")
+    val url = parseEncodedPath("$serverUrl/web/viewer.html")
```
Comment (liudongmiao) [inline plugin/src/main/kotlin/com/firsttimeinforever/intellij/pdf/viewer/jcef/PdfStaticServer.kt:115]: entry point of mozilla's pdf.js

File: plugin/src/main/kotlin/com/firsttimeinforever/intellij/pdf/viewer/jcef/PdfStaticServer.kt [hunk 2]
```diff
@@ -112,11 +112,11 @@ internal class PdfStaticServer : HttpRequestHandler() {
...
     val server = BuiltInServerManager.getInstance()
     return server.addAuthToken(url)
       // `file` would be read via `urlDecoder.path()`, which calls `decodeComponent`
-      .addParameters(mapOf("__reloadSalt" to "$salt", "file" to URLUtil.encodeURIComponent("get-file/${file.url}")))
+      .addParameters(mapOf("__reloadSalt" to "$salt", "file" to "/$uuid/get-file/${URLUtil.encodeURIComponent(file.url)}"))
```
Comment (liudongmiao) [inline plugin/src/main/kotlin/com/firsttimeinforever/intellij/pdf/viewer/jcef/PdfStaticServer.kt:119]: as we use web/viewer.html, so change the file to absolute path.

File: web-view/bootstrap/build.gradle.kts [hunk 3]
```diff
@@ -15,53 +14,45 @@ dependencies {
...
-  if (!file(Paths.get(projectDir.toString(), "node_modules")).exists()) {
-    dependsOn("npm_ci")
+val downloadAndUnzipFile by tasks.registering(Copy::class) {
+  dependsOn(downloadZipFile)
+  from(zipTree(downloadZipFile.get().dest))
+  into(project.layout.buildDirectory)
+  doLast {
+    val viewerHtml = layout.buildDirectory.get().asFile.resolve("web/viewer.html")
+    val modifiedContent = viewerHtml.readText()
+      .replace("(<link rel=\"stylesheet\" href=\"viewer.css\">)".toRegex(), "$1<link rel=\"stylesheet\" href=\"../fixes.css\">")
```
Comment (liudongmiao) [inline web-view/bootstrap/build.gradle.kts:33]: This is the magic 1, add our css after original viewer.css

File: web-view/bootstrap/build.gradle.kts [hunk 4]
```diff
@@ -15,53 +14,45 @@ dependencies {
...
+      .replace("(<script src=\"viewer.m?js\"[^<>]*></script>)".toRegex(), "$1<script src=\"../viewer.js\"></script>")
```
Comment (liudongmiao) [inline web-view/bootstrap/build.gradle.kts:34]: This is the magic 2, add our js after original viewer.js

File: web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/ApplicationFactory.kt [hunk 5]
```diff
@@ -2,22 +2,63 @@ package com.firsttimeinforever.intellij.pdf.viewer.application
...
+      val application = createApplication(window.asDynamic().PDFViewerApplication)
+      startApplication(application)
+    }
+  }
+
+}
+
+@ExperimentalSerializationApi
+fun main() {
+  ApplicationFactory.init()
```
Comment (liudongmiao) [inline web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/ApplicationFactory.kt:63]: point to inject our javascript.

**Comment by liudongmiao** [comment]
This version won't use `pdf_viewer.css`, it may break theme values.
However, there are `viewerCssTheme` and `pageColorsBackground`, `pageColorsForeground`.
`pageColorsXXX` is available since pdf.js v2.14+, in commit https://github.com/mozilla/pdf.js/commit/c8afd6ce8ce2e23aee947f56cc3e44b18415ab8d

File: web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/pdfjs/types/InternalOutline.kt [hunk 6]
```diff
@@ -1,10 +1,11 @@
 package com.firsttimeinforever.intellij.pdf.viewer.application.pdfjs.types
 
 import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.JsonElement
 
 @Serializable
 data class InternalOutline(
-  val dest: String,
+  val dest: JsonElement,
```
Comment (liudongmiao) [inline web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/pdfjs/types/InternalOutline.kt:8]: the dest in outline is {string | Array<any> | null}

**Review by PHPirates — APPROVED** [review]
Great! This solution is a lot cleaner. I'll create a release with this included.

**Review by PHPirates — COMMENTED** [review]
> This version won't use `pdf_viewer.css`, it may break theme values. However, there are `viewerCssTheme` and `pageColorsBackground`, `pageColorsForeground`. `pageColorsXXX` is available since pdf.js v2.14+, in commit [mozilla/pdf.js@c8afd6c](https://github.com/mozilla/pdf.js/commit/c8afd6ce8ce2e23aee947f56cc3e44b18415ab8d)



Wait, as you say, something is not working with the theme. The background color of the sidebar is not correct anymore:
before: 
image

now: 
image

**Comment by liudongmiao** [comment]
@PHPirates it's nothing with `pdf_viewer.css`. However, it's other css changes in pdf viewer.

https://github.com/FirstTimeInForever/intellij-pdf-viewer/blob/fa79e56b99524cef5d27bd0ed2837b255ecb3d48/web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/pdfjs/ThemeUtils.kt#L44-L50

There is no `border-image` in `.pdfViewer .page` since v3.3.122 , see https://github.com/mozilla/pdf.js/commit/685b5866e496345004f4fb5012dd8165906c504d

There is no `box-shadow` in `.thumbnailImage` since v3.6.172, see https://github.com/mozilla/pdf.js/commit/1b6a83da4a0f32b309df5e1fb4143df7970f1915

I'll use the this new way with the old pdf.js, may be a downgrade until we solve the css.

**Comment by liudongmiao** [comment]
I don't see much difference. However, there is scrollbar in new version.

up: pdfjs-dist, with pdf.js 4.6
below: market version
light

dark

**Comment by liudongmiao** [comment]
@PHPirates I have added severa commits, please help to review.

And, I cannot reproduct your sidebar's screenshot.

And, for the sidebar on macos, now, it shows always. However, I think it's acceptable, and Markdown plugin show the scrollbar always too. And, the scrollbar looks like the same as markdown plugin now.

**Comment by liudongmiao** [comment]
> Wait, as you say, something is not working with the theme. The background color of the sidebar is not correct anymore:

@PHPirates Just found the differences. In pdf.js v2.7 and v0.16.1, the background color of sidebar is `rgba(0, 0, 0, 0.1)`
https://github.com/FirstTimeInForever/intellij-pdf-viewer/blob/75b76e6268d81abc9c22cb17ef812a5bc0a452c1/web-view/bootstrap/src/viewer.css#L258-L266

https://github.com/mozilla/pdf.js/blob/v2.7.350/web/viewer.css#L258-L266
However, upstream change the background-color in https://github.com/mozilla/pdf.js/commit/685b5866e496345004f4fb5012dd8165906c504d
then remove the background-color (then inherit background color from body) in https://github.com/mozilla/pdf.js/commit/8ab5476cb82b35dc0407ec76fd0d3223051dd018
both is in pdf.js v3.3.122

However, intellij-pdf-viewer want to change sidebar background color to the same as body background color.
https://github.com/FirstTimeInForever/intellij-pdf-viewer/blob/fa79e56b99524cef5d27bd0ed2837b255ecb3d48/web-view/viewer/src/jsMain/kotlin/com/firsttimeinforever/intellij/pdf/viewer/application/pdfjs/ThemeUtils.kt#L17-L18

However, I don't know why your screenshot don't update the background. Please use DevTools to find it.
You can change the registry `ide.browser.jcef.contextMenu.devTools.enabled` to `true`, then open devtools.
ref: https://plugins.jetbrains.com/docs/intellij/jcef.html#debugging

**Comment by PHPirates** [comment]
Thanks for the explanation! Sorry must have been something wrong on my side, after a gradle clean everything works as expected. Also changing the background color in the PDF Viewer settings work fine (foreground `--main-color` doesn't look like it changes anything, but that was already in 0.16.1 so is not caused by the pdf.js update).

The scrollbar looks good to me, much better than it was!

**Comment by liudongmiao** [comment]
@PHPirates I self-review again, and it could be merged into master now.

For the events in 

https://github.com/FirstTimeInForever/intellij-pdf-viewer/commit/062acd1ef93892b326cdcb503e8b94bd40810d1f#diff-85699ae3492958e0a83d3d768422d1124006a9213eaa31c775de20f22a4755a5L223-L243

it's introduced in https://github.com/FirstTimeInForever/intellij-pdf-viewer/commit/a2d6bddffd04179f4b48d78951606261ac45f8ab, the commit message is to add mouse shortcuts for zoom in/out, however, it's core function was removed as in another commit https://github.com/FirstTimeInForever/intellij-pdf-viewer/commit/ad34116b4d37100700f15cdc7a6389c3bbb4602a

**Comment by PHPirates** [comment]
Great, thanks for checking!

**Pull request merged**